### ISSUE TITLE
resource limits are not respected

### ISSUE THREAD

**Issue by Brice187** [mongodb/mongodb-kubernetes-operator#1057]
**What did you do to encounter the bug?**
```yaml
apiVersion: mongodbcommunity.mongodb.com/v1
kind: MongoDBCommunity
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: >
      {"apiVersion":"mongodbcommunity.mongodb.com/v1","kind":"MongoDBCommunity","metadata":{"annotations":{},"labels":{"argocd.argoproj.io/instance":"example-portal-stage"},"name":"portaldb","namespace":"portal"},"spec":{"members":3,"prometheus":{"passwordSecretRef":{"name":"metrics-endpoint-password"},"username":"prometheus"},"security":{"authentication":{"modes":["SCRAM"]},"tls":{"caConfigMapRef":{"name":"example-tls-ca-configmap"},"certificateKeySecretRef":{"name":"portal-tls-secret"},"enabled":true}},"statefulSet":{"spec":{"template":{"spec":{"containers":[{"name":"mongod","resources":{"limits":{"cpu":"0.1","memory":"250MG"},"requests":{"cpu":"0.1","memory":"200M"}}},{"name":"mongodb-agent","resources":{"limits":{"cpu":"0.1","memory":"250MG"},"requests":{"cpu":"0.1","memory":"200M"}}}]}}}},"type":"ReplicaSet","users":[{"db":"admin","name":"mongoadmin","passwordSecretRef":{"name":"mongoadmin-password"},"roles":[{"db":"admin","name":"clusterAdmin"}],"scramCredentialsSecretName":"mongoadmin-scram"},{"db":"portaldb","name":"portaluser","passwordSecretRef":{"name":"portaluser-password"},"roles":[{"db":"portaldb","name":"readWrite"}],"scramCredentialsSecretName":"portaluser-scram"},{"db":"portaldb","name":"portaluser-ro","passwordSecretRef":{"name":"portaluser-ro-password"},"roles":[{"db":"portaldb","name":"read"}],"scramCredentialsSecretName":"portaluser-ro-scram"}],"version":"5.0.9"}}
    mongodb.com/v1.lastAppliedMongoDBVersion: 5.0.9
    mongodb.com/v1.lastSuccessfulConfiguration: >-
      {"members":3,"type":"ReplicaSet","version":"5.0.9","arbiters":0,"security":{"authentication":{"modes":["SCRAM"],"ignoreUnknownUsers":true},"tls":{"enabled":true,"optional":false,"certificateKeySecretRef":{"name":"portal-tls-secret"},"caConfigMapRef":{"name":"example-tls-ca-configmap"}}},"users":[{"name":"mongoadmin","db":"admin","passwordSecretRef":{"name":"mongoadmin-password","key":""},"roles":[{"db":"admin","name":"clusterAdmin"}],"scramCredentialsSecretName":"mongoadmin-scram","connectionStringSecretName":""},{"name":"portaluser","db":"portaldb","passwordSecretRef":{"name":"portaluser-password","key":""},"roles":[{"db":"portaldb","name":"readWrite"}],"scramCredentialsSecretName":"portaluser-scram","connectionStringSecretName":""},{"name":"portaluser-ro","db":"portaldb","passwordSecretRef":{"name":"portaluser-ro-password","key":""},"roles":[{"db":"portaldb","name":"read"}],"scramCredentialsSecretName":"portaluser-ro-scram","connectionStringSecretName":""}],"statefulSet":{"spec":{}},"additionalMongodConfig":{},"prometheus":{"username":"prometheus","passwordSecretRef":{"name":"metrics-endpoint-password","key":""},"tlsSecretKeyRef":{"name":"","key":""}}}
  creationTimestamp: '2022-05-27T13:53:09Z'
  generation: 13
  labels:
    argocd.argoproj.io/instance: example-portal-stage
  name: portaldb
  namespace: portal
  resourceVersion: '160436902'
  uid: d9094a12-3451-4854-ba26-4302fb55ef03
spec:
  members: 3
  prometheus:
    passwordSecretRef:
      name: metrics-endpoint-password
    username: prometheus
  security:
    authentication:
      ignoreUnknownUsers: true
      modes:
        - SCRAM
    tls:
      caConfigMapRef:
        name: example-tls-ca-configmap
      certificateKeySecretRef:
        name: portal-tls-secret
      enabled: true
  statefulSet:
    spec:
      template:
        spec:
          containers:
            - name: mongod
              resources:
                limits:
                  cpu: '0.1'
                  memory: 250MG
                requests:
                  cpu: '0.1'
                  memory: 200M
            - name: mongodb-agent
              resources:
                limits:
                  cpu: '0.1'
                  memory: 250MG
                requests:
                  cpu: '0.1'
                  memory: 200M
  type: ReplicaSet
  users:
    - db: admin
      name: mongoadmin
      passwordSecretRef:
        name: mongoadmin-password
      roles:
        - db: admin
          name: clusterAdmin
      scramCredentialsSecretName: mongoadmin-scram
    - db: portaldb
      name: portaluser
      passwordSecretRef:
        name: portaluser-password
      roles:
        - db: portaldb
          name: readWrite
      scramCredentialsSecretName: portaluser-scram
    - db: portaldb
      name: portaluser-ro
      passwordSecretRef:
        name: portaluser-ro-password
      roles:
        - db: portaldb
          name: read
      scramCredentialsSecretName: portaluser-ro-scram
  version: 5.0.9
status:
  currentMongoDBMembers: 3
  currentStatefulSetReplicas: 3
  mongoUri: >-
    mongodb://portaldb-0.portaldb-svc.portal.svc.cluster.local:27017,portaldb-1.portaldb-svc.portal.svc.cluster.local:27017,portaldb-2.portaldb-svc.portal.svc.cluster.local:27017/?replicaSet=portaldb
  phase: Running
  version: 5.0.9
```

**What did you expect?**

* container **mongod** has 0.1 CPU / 250MG
* container **mongodb-agent** has 0.1 CPU / 250MG

**What happened instead?**

The Operator generated following **Stateful Set**:

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  creationTimestamp: '2022-07-26T07:18:09Z'
  generation: 1
  labels:
    app: portaldb-svc
  name: portaldb
  namespace: portal
  ownerReferences:
    - apiVersion: mongodbcommunity.mongodb.com/v1
      blockOwnerDeletion: true
      controller: true
      kind: MongoDBCommunity
      name: portaldb
      uid: d9094a12-3451-4854-ba26-4302fb55ef03
  resourceVersion: '160437263'
  uid: bc64a533-555b-4e2a-bf34-474f7bf739c6
spec:
  podManagementPolicy: OrderedReady
  replicas: 3
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: portaldb-svc
  serviceName: portaldb-svc
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: portaldb-svc
    spec:
      containers:
        - command:
            - /bin/sh
            - '-c'
            - >+

              #run post-start hook to handle version changes

              /hooks/version-upgrade


              # wait for config and keyfile to be created by the agent
               while ! [ -f /data/automation-mongod.conf -a -f /var/lib/mongodb-mms-automation/authentication/keyfile ]; do sleep 3 ; done ; sleep 2 ;

              # with mongod configured to append logs, we need to provide them
              to stdout as

              # mongod does not write to stdout and a log file

              tail -F /var/log/mongodb-mms-automation/mongodb.log > /dev/stdout
              &


              # start mongod with this configuration

              exec mongod -f /data/automation-mongod.conf;

          env:
            - name: AGENT_STATUS_FILEPATH
              value: /healthstatus/agent-health-status.json
          image: 'myregistry.mynet/mongo:5.0.9'
          imagePullPolicy: IfNotPresent
          name: mongod
          resources:
            limits:
              cpu: '1'
              memory: 500M
            requests:
              cpu: 500m
              memory: 400M
          securityContext:
            runAsNonRoot: true
            runAsUser: 2000
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          volumeMounts:
            - mountPath: /data
              name: data-volume
            - mountPath: /healthstatus
              name: healthstatus
            - mountPath: /hooks
              name: hooks
            - mountPath: /var/log/mongodb-mms-automation
              name: logs-volume
            - mountPath: /var/lib/mongodb-mms-automation/authentication
              name: portaldb-keyfile
            - mountPath: /var/lib/tls/ca/
              name: tls-ca
              readOnly: true
            - mountPath: /var/lib/tls/server/
              name: tls-secret
              readOnly: true
        - command:
            - /bin/bash
            - '-c'
            - >-
              current_uid=$(id -u)

              AGENT_API_KEY="$(cat
              /mongodb-automation/agent-api-key/agentApiKey)"

              declare -r current_uid

              if ! grep -q "${current_uid}" /etc/passwd ; then

              sed -e "s/^mongodb:/builder:/" /etc/passwd > /tmp/passwd

              echo "mongodb:x:$(id -u):$(id -g):,,,:/:/bin/bash" >> /tmp/passwd

              export NSS_WRAPPER_PASSWD=/tmp/passwd

              export LD_PRELOAD=libnss_wrapper.so

              export NSS_WRAPPER_GROUP=/etc/group

              fi

              agent/mongodb-agent
              -healthCheckFilePath=/var/log/mongodb-mms-automation/healthstatus/agent-health-status.json
              -serveStatusPort=5000
              -cluster=/var/lib/automation/config/cluster-config.json
              -skipMongoStart -noDaemonize -useLocalMongoDbTools
          env:
            - name: AGENT_STATUS_FILEPATH
              value: >-
                /var/log/mongodb-mms-automation/healthstatus/agent-health-status.json
            - name: AUTOMATION_CONFIG_MAP
              value: portaldb-config
            - name: HEADLESS_AGENT
              value: 'true'
            - name: POD_NAMESPACE
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: metadata.namespace
          image: 'myregistry.mynet/mongodb/mongodb-agent:11.12.0.7388-1'
          imagePullPolicy: Always
          name: mongodb-agent
          readinessProbe:
            exec:
              command:
                - /opt/scripts/readinessprobe
            failureThreshold: 40
            initialDelaySeconds: 5
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 1
          resources:
            limits:
              cpu: '1'
              memory: 500M
            requests:
              cpu: 500m
              memory: 400M
          securityContext:
            runAsNonRoot: true
            runAsUser: 2000
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          volumeMounts:
            - mountPath: /opt/scripts
              name: agent-scripts
            - mountPath: /var/lib/automation/config
              name: automation-config
              readOnly: true
            - mountPath: /data
              name: data-volume
            - mountPath: /var/log/mongodb-mms-automation/healthstatus
              name: healthstatus
            - mountPath: /var/log/mongodb-mms-automation
              name: logs-volume
            - mountPath: /var/lib/mongodb-mms-automation/authentication
              name: portaldb-keyfile
            - mountPath: /var/lib/tls/ca/
              name: tls-ca
              readOnly: true
            - mountPath: /var/lib/tls/server/
              name: tls-secret
              readOnly: true
      dnsPolicy: ClusterFirst
      initContainers:
        - command:
            - cp
            - version-upgrade-hook
            - /hooks/version-upgrade
          image: >-
            myregistry.mynet/mongodb/mongodb-kubernetes-operator-version-upgrade-post-start-hook:1.0.4
          imagePullPolicy: Always
          name: mongod-posthook
          resources: {}
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          volumeMounts:
            - mountPath: /hooks
              name: hooks
        - command:
            - cp
            - /probes/readinessprobe
            - /opt/scripts/readinessprobe
          image: 'myregistry.mynet/mongodb/mongodb-kubernetes-readinessprobe:1.0.9'
          imagePullPolicy: Always
          name: mongodb-agent-readinessprobe
          resources: {}
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          volumeMounts:
            - mountPath: /opt/scripts
              name: agent-scripts
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext:
        fsGroup: 2000
      serviceAccount: mongodb-database
      serviceAccountName: mongodb-database
      terminationGracePeriodSeconds: 30
      volumes:
        - emptyDir: {}
          name: agent-scripts
        - name: automation-config
          secret:
            defaultMode: 416
            secretName: portaldb-config
        - emptyDir: {}
          name: healthstatus
        - emptyDir: {}
          name: hooks
        - emptyDir: {}
          name: portaldb-keyfile
        - name: tls-ca
          secret:
            defaultMode: 416
            secretName: portaldb-ca-certificate
        - name: tls-secret
          secret:
            defaultMode: 416
            secretName: portaldb-server-certificate-key
  updateStrategy:
    type: RollingUpdate
  volumeClaimTemplates:
    - apiVersion: v1
      kind: PersistentVolumeClaim
      metadata:
        creationTimestamp: null
        name: data-volume
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 10G
        volumeMode: Filesystem
      status:
        phase: Pending
    - apiVersion: v1
      kind: PersistentVolumeClaim
      metadata:
        creationTimestamp: null
        name: logs-volume
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 2G
        volumeMode: Filesystem
      status:
        phase: Pending
status:
  availableReplicas: 3
  collisionCount: 0
  currentReplicas: 3
  currentRevision: portaldb-d9665766
  observedGeneration: 1
  readyReplicas: 3
  replicas: 3
  updateRevision: portaldb-d9665766
  updatedReplicas: 3
```

* container **mongod** has 1 CPU / 500MG
* container **mongodb-agent** has 1 CPU / 500MG

**Operator Information**
 - Operator Version: 0.7.4 (HelmChart)
 - MongoDB Image used: 5.0.9

**Kubernetes Cluster Information**
 - k3s
 - 1.24.3
 - internal

**Additional context**

Copied example Code from GH

**Comment by D3DeFi** [comment]
Hello,

This doesn't work neither when creating nor when updating MongoDBCommunity resource as controller always seems to force defaults, which are too much for small Kubernetes clusters.

Excuse the verbosity (brain dump), I don't have experience with Go or programming operators on my side, but I was able to perhaps trace this problem:

Invoking this function before `statefulset.CreateOrUpdate()` should build StatefulSet specification: 
https://github.com/mongodb/mongodb-kubernetes-operator/blob/28cb3e04d37aaab56489d0f883e55e84c88fed4e/controllers/replica_set_controller.go#L500

The problem is, that it never seems to build PodTemplateSpec correctly with provided resource requests/limits as it calls:
```go
commonModification := construct.BuildMongoDBReplicaSetStatefulSetModificationFunction(&mdb, mdb)
```

which in turn invokes `mongodbContainer()`  and `mongodbAgentContainer()` functions when building StatefulSet specification where both always enforce default resource requirements via `resourcerequirements.Defaults()` at:

https://github.com/mongodb/mongodb-kubernetes-operator/blob/28cb3e04d37aaab56489d0f883e55e84c88fed4e/controllers/construct/mongodbstatefulset.go#L179-L180

https://github.com/mongodb/mongodb-kubernetes-operator/blob/28cb3e04d37aaab56489d0f883e55e84c88fed4e/controllers/construct/mongodbstatefulset.go#L316

I think the fix would be something like this in semi-pseudo Go? Can someone with actual Go experience confirm?

```go
rr := additionalMongoDBConfig.GetResourceRequirements()  # needs to be implemented
...
	return container.Apply(
                ...
		container.WithResourceRequirements(rr || resourcerequirements.Defaults()),
```

**Comment by D3DeFi** [comment]
This no longer seems to be an issue for me with 0.7.6 operator

**Comment by Jay-Funk** [comment]
I am still seeing the same issue with 0.7.6

Operator:
``` yaml
resources:
                limits:
                  cpu: "0.5"
                  memory: 2G
                requests:
                  cpu: "0.2"
                  memory: 1G
```

Statefulset:
``` yaml
resources:
                limits:
                  cpu: "1"
                  memory: 500M
                requests:
                  cpu: "500m"
                  memory: 400M
```

**Comment by slaskawi** [comment]
I quickly checked the code and it seems we more tests around this now. Indeed, it was fixed in 0.7.6: https://github.com/mongodb/mongodb-kubernetes-operator/pull/287
